var/spack/repos/builtin/packages/axom/package.py

```diff
@@ -410,20 +414,34 @@ def hostconfig(self, spec, prefix):
             cfg.write(cmake_cache_entry("SHROUD_EXECUTABLE",
                                         pjoin(shroud_bin_dir, "shroud")))
 
-        if "uncrustify" in spec:
-            uncrustify_bin_dir = get_spec_path(spec, "uncrustify",
-                                               path_replacements,
-                                               use_bin=True)
-            cfg.write(cmake_cache_entry("UNCRUSTIFY_EXECUTABLE",
-                                        pjoin(uncrustify_bin_dir,
-                                              "uncrustify")))
-
         if "cppcheck" in spec:
             cppcheck_bin_dir = get_spec_path(spec, "cppcheck",
                                              path_replacements, use_bin=True)
             cfg.write(cmake_cache_entry("CPPCHECK_EXECUTABLE",
                                         pjoin(cppcheck_bin_dir, "cppcheck")))
 
+        # Only turn on clangformat support if devtools is on
+        if "+devtools" in spec:
+            cf_paths = []
+            lc_clangpath = "/usr/tce/packages/clang/clang-10.0.0"
+            cf_paths.append(pjoin(lc_clangpath, "bin/clang-format"))
+            cf_paths.append("/usr/bin/clang-format-10")
+            cf_paths.append("/usr/bin/clang-format")
```

Comment:
What's the deal with all of these system paths?

Reply:
We support multiple systems and since we want to tie our clang-format tool to a version this seemed to be the easiest/cleanest way we can do that.  I am open to suggestions but I would prefer this not being copy-pasted in a lot of places.  I did guard it with a variant that is only used by developers.

Reply:
What if we instead add a dependency on whatever package provides this executable (LLVM?). Then you can add it as an external package on those systems, or let Spack build it for you.

Reply:
ClangFormat comes with the clang compiler, as far as I know. I would prefer not to have for example gcc depend on clang.

Reply:
Well luckily the variant is off by default. I think the right thing to do here is to add a dependency on `llvm+clang`. You can always add it as an external package if you don't want to build it. But then this package will work for any user on any system.

Reply:
Sounds good.  I will add an issue to Axom and do that in a follow up PR, if thats okay.

Reply:
Hmm, I would prefer if you either:

1. Did it now, should be easy, or
2. Removed the logic for these system paths and do it in a follow-up PR

Reply:
The problem is the clang-format version has to be very specific (10.0.0) due to differences in formatting between versions.  This version is completely different from any other dependency/package. I am concerned about the unnecessary constraints this puts on concretization for one executable.  This maybe unfounded.

Reply:
I would try it out and see. Hopefully the new concretizer will solve any concretization issues you encounter. And that's another reason why it's good that the variant is off by default.

Reply:
these paths are checked and if valid added -- that seems like a safe compromise. 
it will help folks that are building on these systems and shouldn't undermine anyone else. 

depending on llvm+clang seems really scary here?  Is spack's compiler logic totally orthogonalized such that depending on llvm here (just for clang-format) won't entangle things when we use other compilers?

is there a way to create a clang-format package - that we could set external? That might be safer.

Reply:
I could just remove all this logic, but we share this package with the development version of Axom (and others) and I have gone through a lot of effort to make it so that there are no custom changes in this file.  I am hesitant to start down that path yet again.

Reply:
I think @adamjstewart is right that the correct way to do this is with an explicit dependency.

@cyrush @white238, if you use an external in your configuration for `llvm+clang`, then the additional concretization constraints will be very minor, and go away entirely with the new concretizer (because that will allow llvm+clang to be automatically built with a different compiler.

The compiler and depencency models are sufficiently orthogonal that this should not cause any issue. We are working on a refactor of the compiler model as you know, and I've added this specific use-case to the set of constraints for that refactor.

Reply:
> is there a way to create a clang-format package - that we could set external? That might be safer.

In the past, I believe we had a lot of LLVM extensions as separate packages, but it seems like the current push is to combine all of them in the base `llvm` package. Another example is `llvm-openmp`, which I rely on for a lot of packages on macOS where the system Apple Clang doesn't have OpenMP support, but I don't want to build all of LLVM everytime.

@naromero77 @trws may have something to say about this.

Reply:
Yes, the llvm sub-packages are currently all part of one package.  This is partially as a result of that being how they are built, and partially because the monorepo is only really meant to be used this way.  We might be able to set up a clang-format package, but I'm not sure it would be all that meaningful other than in a binary context with the main llvm package broken up into component parts as well.  Something worth looking into maybe, but it's not a simple issue when building from source (or at least wasn't when I last tried, which was admittedly ~4 major versions ago).